tools/tmt/README.md

```diff
@@ -0,0 +1,283 @@
+
+Where
+
+  * `X` stands for the type map format version
+  * `CONFIGURATION` is either `Debug` or `Release`
+  * `ARCHITECTURE` is one of: `ARM`, `ARM64`, `X86`, `X86_64`
+  * `KIND` is either `managed` or `java`
+
+If regular expression type name filtering is not enabled, the utility
+will not print the map contents to the console.  This is because even
+small applications can contain thousands of entries and the whole map
+content is better written to a file instead of producing voluminous
+output on the screen.
+
+## Filtering
+
+### Type name filtering
+
+Passing a regular expression to the utility causes it to match type
+names (both Managed and Java for **each** loaded entry, regardless on
+whether the current map being processed is java-to-managed or
+managed-to-java) against it and report the matches to the console.
+
+Also, in this case the files are not generated by the default, which
+can be changed by using the `-g` option.  This is because filtering
+mode is mostly useful for "quick look" at the maps and the full output
+will likely be of no interest in this instance.
+
+### Architecture filtering
+
+By default the utility reads `libxamarin-app.so` for any architecture
+it finds in the location indicated on command line.  The `-a` option
+can be used to limit the set of architectures by specifying a
+comma-separated list of architecture names from the following set
+(architecture names are case-insensitive): `ARM`, `ARM64`, `X86` and
+`X86_64`
+
+# Command line options
+
+```
+Usage: tmt [OPTIONS] <FILE.apk|FILE.aab|libxamarin-app.so|PROJECT_DIR> [FILTER_
+REGEX]
+
+OPTIONS are:
+
+  -j, --only-java            Process only the java-to-managed map
+  -m, --only-managed         Process only the managed-to-java map
+  -s, --short-report         Omit some map details from the report (e.g. MVID
+                               and TokenID from managed-to-java map)
+  -o, --output-directory=DIR Write the report files in the DIR directory
+                               instead of the current one
+  -a, --arch=ARCH_LIST       Limit reporting only to the specified
+                               architectures. ARCH_LIST is a comma-separated
+                               list of architectures (one of, case-insensitive:
+                               ARM, ARM64, X86, X86_64)
+  -1, --only-first           Process only the first shared library from the APK/
+                               AAB archive or a project directory. Architecture
+                               filter is ignored in this case. This is the
+                               default action when regex filtering is used,
+                               otherwise it defaults to `false`
+  -g, --generate-files       Generate report files. If regex filtering is used,
+                               this setting defaults to `false`, it is `true`
+                               otherwise.
+
+  -v, --verbose              Show debug messages
+```
+
+## Sample invocations
```

Comment:
Part of the request for sample invocations was to see what output was generated, which also isn't described anywhere.  (This is why I built the branch, on the hopes I could see what it was like and consider updating the commit message/`README.md`… then found that the `tmt` script didn't exist.)

Now that I have the output, I see that it's a "tab-delimited format":

```
Java Type Name	Managed Type Name	Generic?	Duplicate?
android/accessibilityservice/AccessibilityButtonController	Android.AccessibilityServices.AccessibilityButtonController, Mono.Android		
```

This is good, as far as it goes, but there is no good description of what the "Generic?" column is for, and it doesn't appear to be used in the output for `samples/HelloWorld`.

The `Duplicate?` column is presumably for when the same Java type is mapped to multiple C# types, a'la 66f72064e00499e3c87c57def3f7ca2cad07e1be, but even then I'm not quite sure how to interpret it?

```
android/accessibilityservice/AccessibilityGestureEvent	Android.AccessibilityServices.AccessibilityGestureEvent, Mono.Android		
android/accessibilityservice/AccessibilityService	Android.AccessibilityServices.AccessibilityService, Mono.Android		aliased
```

`android/accessibilityservice/AccessibilityService` has an `aliased` entry, but what's the *non*-aliased entry? Or does `aliased` mean that there is more than one C# type, *period*, and indicates nothing about *preferred* mapping?

Looking at the managed-to-Java mapping, this "no preference" interpretation appears correct, as only one entry is listed Java-to-managed (above):

```
Managed Type Name	Java Type Name	Generic?	Duplicate?
Android.AccessibilityServices.AccessibilityService, Mono.Android	android/accessibilityservice/AccessibilityService		
…
Android.AccessibilityServices.AccessibilityServiceInvoker, Mono.Android	android/accessibilityservice/AccessibilityService		
```

TL;DR: these files are also part of the usage pattern of this app.  They should be likewise described.  I'm happy to help with that description, but I can't easily fix the "missing `tmt` script" problem.

Reply:
Oh, and of course the Release format differs, so that'll need an explanation as well:

```
Java Type Name	Managed Type Name	MVID	Token ID
…
java/lang/Object	Java.Lang.Object, Mono.Android	dc97aced-a52d-4110-bb03-b57c6368c440	33554768 (0x02000150)
```

Reply:
I wonder if these headers should use `-` instead of space ` ` within the description, to make it easier to visually distinguish the tabs, e.g.

```
Java-Type-Name	Managed-Type-Name	MVID	Token-ID
```

Reply:
> Part of the request for sample invocations was to see what output was generated, which also isn't described anywhere. (This is why I built the branch, on the hopes I could see what it was like and consider updating the commit message/`README.md`… then found that the `tmt` script didn't exist.)
> 
Not sure why your `dotnet` doesn't produce the `tmt` binary, I would assume some sort of consistency when building the same project with the same tool...

> ```
> Java Type Name	Managed Type Name	Generic?	Duplicate?
> android/accessibilityservice/AccessibilityButtonController	Android.AccessibilityServices.AccessibilityButtonController, Mono.Android		
> ```
> 
> This is good, as far as it goes, but there is no good description of what the "Generic?" column is for, and it doesn't appear to be used in the output for `samples/HelloWorld`.
> 
Where would I put the description? The column headers should be short and I think that, considering we're looking at the listing of C# and Java types, the word would be clear to mean a `Generic type`. As for the column being empty in `samples/HelloWorld` output... well, can't do much about it, can I? :)

> The `Duplicate?` column is presumably for when the same Java type is mapped to multiple C# types, a'la [66f7206](https://github.com/xamarin/xamarin-android/commit/66f72064e00499e3c87c57def3f7ca2cad07e1be), but even then I'm not quite sure how to interpret it?
> 
There's not much to interpret here - it's what you presumed originally. Again, the assumption is that the person using the tool knows what typemaps are and how to interpret the word in the context.

> ```
> android/accessibilityservice/AccessibilityGestureEvent	Android.AccessibilityServices.AccessibilityGestureEvent, Mono.Android		
> android/accessibilityservice/AccessibilityService	Android.AccessibilityServices.AccessibilityService, Mono.Android		aliased
> ```
> 
> `android/accessibilityservice/AccessibilityService` has an `aliased` entry, but what's the _non_-aliased entry? Or does `aliased` mean that there is more than one C# type, _period_, and indicates nothing about _preferred_ mapping?
> 
Jon, what are we trying to achieve here? Do you want the output of the tool to explain how typemaps work and what they are? I hardly think this is the correct place to do it. The format and the workings are documented elsewhere and I **really** think that the utility is not for general use by every developer using `Xamarin.Android`, but for forensic examination by those who need to dive deep enough to have to learn what typemaps are (i.e. mostly the XA team) - and if somebody goes that deep to the details of what we put in the APK, I'm sure they will know what "generic", "duplicate" and "aliased" are. If you don't like these terms, please tell me what to put there as I think they're clear enough.

> Looking at the managed-to-Java mapping, this "no preference" interpretation appears correct, as only one entry is listed Java-to-managed (above):
> 
> ```
> Managed Type Name	Java Type Name	Generic?	Duplicate?
> Android.AccessibilityServices.AccessibilityService, Mono.Android	android/accessibilityservice/AccessibilityService		
> …
> Android.AccessibilityServices.AccessibilityServiceInvoker, Mono.Android	android/accessibilityservice/AccessibilityService		
> ```
> 
> TL;DR: these files are also part of the usage pattern of this app. They should be likewise described. I'm happy to help with that description, but I can't easily fix the "missing `tmt` script" problem.

I can't fix that either as I don't know why `dotnet` doesn't produce the binary on your system but creates it on mine. I will instead provide an actual script which will work in both cases.

Reply:
> Do you want the output of the tool to explain how typemaps work and what they are? 

No, not the *output of the tool*;  The `README.md` of the tool.  Which is why this comment thread started on the `README.md`.

And again, I'd be happy to write up that description, I just wasn't sure what some of the values were supposed to mean.

For example, my initial Hello World build did no linking, and thus was *huge* (yay?), with 6062 Java-to-managed entries and 5376 Managed-to-Java entries.

This should be ~everything in `Mono.Android.dll`, right? (Perhaps not?  There are 10848 TypeDefinitions within `Mono.Android.dll`, though not every typedef is a distinct binding, so "half-ish" might be a sensible value.)

So why don't any of these entries appear to produce output in the `Is-Generic-Type?` column?  Or am I missing something?  I thought we needed the `generic` column for e.g. `Android.Runtime.JavaList<T>`?

…which oddly isn't present?!

```
% grep JavaList x/typemap-v1-Debug-X86_64-managed.txt
Android.Runtime.JavaList, Mono.Android	java/util/ArrayList		duplicate entry
```

There *is* a generic type present, but it doesn't have anything in the `Is-Generic-Type?` column:

```
% grep JavaArray x/typemap-v1-Debug-X86_64-managed.txt
Android.Runtime.JavaArray`1, Mono.Android	mono/android/runtime/JavaArray		
```

Why doesn't `JavaArray<T>` have a value for `Is-Generic-Type`?

For that matter, why is `JavaArray<T>` the only generic type in here?

```
% grep '`' x/typemap-v1-Debug-X86_64-managed.txt
Android.Runtime.JavaArray`1, Mono.Android	mono/android/runtime/JavaArray		
```

At this point I start wondering if something is wrong with my setup…?